dtfs-central-api/src/services/state-machines/utilisation-report/event-handlers/add-a-payment/add-a-payment.event-handler.test.ts

```diff
@@ -0,0 +1,278 @@
+  PaymentEntity,
+  UTILISATION_REPORT_RECONCILIATION_STATUS,
+  UtilisationReportEntity,
+  UtilisationReportEntityMockBuilder,
+} from '@ukef/dtfs2-common';
+import { handleUtilisationReportAddAPaymentEvent } from './add-a-payment.event-handler';
+import { NewPaymentDetails } from '../../../../../types/utilisation-reports';
+import { FeeRecordStateMachine } from '../../../fee-record/fee-record.state-machine';
+import { feeRecordsMatchAttachedPayments } from './helpers';
+
+jest.mock('./helpers');
+
+describe('handleUtilisationReportAddAPaymentEvent', () => {
+  const tfmUserId = 'abc123';
+  const requestSource: DbRequestSource = {
+    platform: 'TFM',
+    userId: tfmUserId,
+  };
+
+  const mockEntityManager = {
+    save: jest.fn(),
+  } as unknown as EntityManager;
+
+  const paymentCurrency: Currency = 'GBP';
+  const paymentDetails: NewPaymentDetails = {
+    currency: paymentCurrency,
+    amountReceived: 100,
+    dateReceived: new Date(),
+    paymentReference: 'A payment reference',
+  };
+
+  const aListOfFeeRecordsForReport = (report: UtilisationReportEntity): FeeRecordEntity[] => [
+    FeeRecordEntityMockBuilder.forReport(report).withId(1).withPaymentCurrency(paymentCurrency).build(),
+    FeeRecordEntityMockBuilder.forReport(report).withId(2).withPaymentCurrency(paymentCurrency).build(),
+  ];
+
+  const getFeeRecordStateMachinesForFeeRecords = (feeRecords: FeeRecordEntity[]) =>
+    feeRecords.map((feeRecord) => FeeRecordStateMachine.forFeeRecord(feeRecord));
+
+  const getEventHandlerSpiesForFeeRecordStateMachines = (feeRecordStateMachines: FeeRecordStateMachine[], feeRecords: FeeRecordEntity[]) =>
+    feeRecordStateMachines.map((stateMachine, index) => jest.spyOn(stateMachine, 'handleEvent').mockResolvedValue(feeRecords[index]));
+
+  const mockFeeRecordStateMachineConstructorImplementation = (feeRecords: FeeRecordEntity[], feeRecordStateMachines: FeeRecordStateMachine[]) => {
+    jest.spyOn(FeeRecordStateMachine, 'forFeeRecord').mockImplementation((feeRecord) => {
+      const indexOfMatchingFeeRecord = feeRecords.findIndex(({ id }) => id === feeRecord.id);
+      if (indexOfMatchingFeeRecord === -1) {
+        throw new Error('Failed to find fee record with matching id');
+      }
+      return feeRecordStateMachines[indexOfMatchingFeeRecord];
+    });
+  };
+
+  beforeEach(() => {
+    jest.mocked(feeRecordsMatchAttachedPayments).mockResolvedValue(false);
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it('creates and saves the new payment entity using the supplied entity manager', async () => {
+    // Arrange
+    const utilisationReport = UtilisationReportEntityMockBuilder.forStatus('PENDING_RECONCILIATION').build();
+
+    const feeRecords = aListOfFeeRecordsForReport(utilisationReport);
+    const feeRecordStateMatchines = getFeeRecordStateMachinesForFeeRecords(feeRecords);
+    mockFeeRecordStateMachineConstructorImplementation(feeRecords, feeRecordStateMatchines);
```

Comment:
I didn't do this originally because I was concerned about how tricky it would be to mock the state machine, but it's easy enough to just include the following at the bottom of the file

```typescript
function aMockFeeRecordStateMachine(): FeeRecordStateMachine {
  return {
    handleEvent: jest.fn(),
  } as unknown as FeeRecordStateMachine;
}
```